be/src/exec/vectorized/partition/partition_hash_map.h

```diff
@@ -0,0 +1,288 @@
+template <PhmapSeed seed>
+using FixedSize16SlicePartitionHashMap =
+        phmap::flat_hash_map<SliceKey16, PartitionChunks, FixedSizeSliceKeyHash<SliceKey16, seed>>;
+
+struct PartitionHashMapBase {
+    const int32_t chunk_size;
+
+    PartitionHashMapBase(int32_t chunk_size) : chunk_size(chunk_size) {}
+
+protected:
+    void alloc_new_buffer(PartitionChunks& value, const ChunkPtr& chunk) {
+        static size_t reserve_size = 1;
+        // Cause we don't know the cardinality of the partition columns, so we
+        // shouldn't reserve too much space for the buffered chunk
+        // Now the reserve size is set to 1, advanced mechanism may be introduced later
+        ChunkPtr new_chunk = chunk->clone_empty_with_slot(reserve_size);
+        value.select_indexes.clear();
+        value.select_indexes.reserve(reserve_size);
+        value.chunks.push_back(std::move(new_chunk));
+        value.remain_size = chunk_size;
+    }
+
+    void flush(PartitionChunks& value, const ChunkPtr& chunk) {
+        if (!value.chunks.empty() && !value.select_indexes.empty()) {
+            value.chunks.back()->append_selective(*chunk, value.select_indexes.data(), 0, value.select_indexes.size());
+            value.select_indexes.clear();
+            value.remain_size = chunk_size - value.chunks.back()->num_rows();
+        }
+    }
+
+    // Append chunk to the hash_map for one non-nullable key
+    // Each key mapped to a PartitionChunks which holds an array of chunks
+    // chunk will be divided into multiply parts by partition key, and each parts will be append to the
+    // last chunk of PartitionChunks.chunks. New chunk will be allocated if the last chunk reaches its capacity
+    // @key_loader used to load key for number type or slice type
+    template <typename HashMap, typename KeyLoader, typename KeyAllocator>
+    void append_chunk_for_one_key(HashMap& hash_map, ChunkPtr chunk, KeyLoader&& key_loader,
+                                  KeyAllocator&& key_allocator) {
+        const auto size = chunk->num_rows();
+        for (uint32_t i = 0; i < size; i++) {
+            const auto& key = key_loader(i);
+            auto iter = hash_map.lazy_emplace(
+                    key, [&](const auto& ctor) { return ctor(key_allocator(key), PartitionChunks()); });
+            auto& value = iter->second;
+            if (value.chunks.empty() || value.remain_size <= 0) {
+                if (!value.chunks.empty()) {
+                    value.chunks.back()->append_selective(*chunk, value.select_indexes.data(), 0,
+                                                          value.select_indexes.size());
+                }
+
+                alloc_new_buffer(value, chunk);
+            }
+            value.select_indexes.push_back(i);
+            value.remain_size--;
+        }
+
+        for (auto& [key, value] : hash_map) {
+            flush(value, chunk);
```

Comment:
In extreme scenarios, such a Chunk may be split into 4096 lines, and then written to other 4096 Chunks, the memory locality is very poor, and the performance may be poor. Can test it later

Reply:
Especially in the case of multiple columns

Reply:
Is it possible to not copy non-Partition columns ?

Reply:
Ok, I'll test it later. And partitioner is a dependent component which is designed to divide chunks into multiply parts, so in this way, it can't copy only partition columns